pkg/controller/prowjob/prowjob-controller.go

```diff
@@ -0,0 +1,147 @@
+	SuccessSynced = "Synced"
+
+	// MessageResourceSynced is the message used for an Event fired when a ProwJob is synced successfully
+	MessageResourceSynced = "ProwJob synced successfully"
+)
+
+// ProwJobController is the controller implementation for ProwJob resources
+type ProwJobController struct {
+	controller.Controller
+	prowJobLister lister.ProwJobLister
+	// ctx is the passed-down global context. It's used and passed
+	// down to all API client calls as to make sure all in-flight calls get
+	// cancelled if the main context is
+	ctx context.Context
+	// controllerCancel is the components global cancelFunc. It's used to
+	// cancel the global context, stop the leader election and subsequently
+	// initiate a shut down of all control loops
+	controllerCancel context.CancelFunc
+	// Recorder is an event recorder for recording Event resources to the
+	// Kubernetes API.
+	recorder record.EventRecorder
+}
+
+// NewProwJobController returns a new ProwJob controller
+func NewProwJobController(
+	controllerContext context.Context,
+	controllerCancel context.CancelFunc,
+	kubeClientset kubernetes.Interface,
+	prowJobInformer informer.ProwJobInformer,
+	prowJobNamespace string) *controller.Controller {
+
+	utilruntime.Must(prowscheme.AddToScheme(scheme.Scheme))
+
+	// Create event broadcaster
+	klog.V(4).Info("Creating event broadcaster")
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartStructuredLogging(0)
+	eventBroadcaster.StartRecordingToSink(&typedcorev1.EventSinkImpl{Interface: kubeClientset.CoreV1().Events(prowJobNamespace)})
+	recorder := eventBroadcaster.NewRecorder(scheme.Scheme, corev1.EventSource{Component: prowJobControllerAgentName})
+
+	prowJobController := &ProwJobController{
+		prowJobLister:    prowJobInformer.Lister(),
+		ctx:              controllerContext,
+		controllerCancel: controllerCancel,
+		recorder:         recorder,
+	}
+
+	c := controller.NewController(
+		[]cache.InformerSynced{prowJobInformer.Informer().HasSynced},
+		prowJobController,
+		prowJobControllerAgentName,
+		reflect.TypeOf(&v1.ProwJob{}),
+	)
+
+	prowJobFilter := func(obj interface{}) bool {
+		if prowJob, ok := obj.(*v1.ProwJob); ok {
+			if _, ok := prowJob.Labels["release.openshift.io/verify"]; ok {
+				return true
+			}
+		}
+		if tombstone, ok := obj.(cache.DeletedFinalStateUnknown); ok {
+			if prowJob, ok := tombstone.Obj.(*v1.ProwJob); ok {
+				if _, ok = prowJob.Labels["release.openshift.io/verify"]; ok {
+					return true
+				}
+			}
+		}
+		return false
+	}
+
+	prowJobInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: prowJobFilter,
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc: c.Enqueue,
+			UpdateFunc: func(old interface{}, new interface{}) {
+				oldProwJob, _ := old.(*v1.ProwJob)
+				newProwJob, _ := new.(*v1.ProwJob)
+
+				// Don't process resync or objects that are marked for deletion
```

Comment:
This is really weird. Why?

Reply:
Im assuming that it's an optimization to not waste cycles on identical objects and anything that's actually being deleted.  It seemed logical to me.  Why do you thing it's weird?

Reply:
Did this get copied from somewhere? I have not seen this pattern in the past.

Reply:
Yeah, the same folks that wrote the shared controller logic, use this same pattern to process their update events: https://github.com/openshift/cloud-network-config-controller/blob/224020ca9f043e8023e68dcc133d9a957025d6d1/pkg/controller/secret/secret_controller.go#L68-L91